neurolib/models/multimodel/builder/thalamus.py

```diff
@@ -0,0 +1,442 @@
+        )
+
+        # channel dynamics: T-type and rectifier current
+        d_h_T = (self._h_inf_T(voltage) - h_T) / self._tau_h_T(voltage)
+        d_m_h1 = (
+            (self._m_inf_h(voltage) * (1.0 - m_h2) - m_h1) / self._tau_m_h(voltage)
+            - self.params["k3"] * self._P_h(ca_conc) * m_h1
+            + self.params["k4"] * m_h2
+        )
+        d_m_h2 = self.params["k3"] * self._P_h(ca_conc) * m_h1 - self.params["k4"] * m_h2
+
+        # synaptic dynamics
+        d_syn_ext = dsyn_ext
+        d_syn_inh = dsyn_inh
+        d_dsyn_ext = (
+            self.params["gamma_e"] ** 2
+            * (
+                coupling_variables["node_exc_exc"]
+                + coupling_variables["network_exc_exc"]
+                + system_input(self.noise_input_idx[0])
+                - syn_ext
+            )
+            - 2 * self.params["gamma_e"] * dsyn_ext
+        )
+        d_dsyn_inh = (
+            self.params["gamma_r"] ** 2 * (coupling_variables["node_exc_inh"] - syn_inh)
+            - 2 * self.params["gamma_r"] * dsyn_inh
+        )
+        # firing rate as dummy dynamical variable with infinitely fast
+        # fixed-point dynamics
+        firing_rate_now = self._get_firing_rate(voltage)
+        d_firing_rate = -self.params["lambda"] * (firing_rate - firing_rate_now)
+
+        return [
+            d_voltage,
+            d_ca_conc,
+            d_h_T,
+            d_m_h1,
+            d_m_h2,
+            d_syn_ext,
+            d_syn_inh,
+            d_dsyn_ext,
+            d_dsyn_inh,
+            d_firing_rate,
+        ]
+
+
+class ThalamicReticularPopulation(ThalamicMass):
+    """
+    Inhibitory mass representing thalamic reticular nuclei neurons in the
+    thalamus.
+    """
+
+    name = "Thalamic reticular nuclei mass"
+    label = "TRN"
+    mass_type = INH
+
+    num_state_variables = 7
+    num_noise_variables = 1
+    coupling_variables = {6: f"q_mean_{INH}"}
+    required_couplings = ["node_inh_exc", "node_inh_inh", "network_inh_exc"]
+    state_variable_names = [
+        "V",
+        "h_T",
+        "s_e",
+        "s_i",
+        "ds_e",
+        "ds_i",
+        "q_mean",
+    ]
+    required_params = [
+        "tau",
+        "Q_max",
+        "theta",
+        "sigma",
+        "C1",
+        "C_m",
+        "gamma_e",
+        "gamma_r",
+        "g_L",
+        "g_GABA",
+        "g_AMPA",
+        "g_LK",
+        "g_T",
+        "E_AMPA",
+        "E_GABA",
+        "E_L",
+        "E_K",
+        "E_Ca",
+        "ext_current",
+        "lambda",
+    ]
+
+    def __init__(self, params=None):
+        super().__init__(params=params or DEFAULT_PARAMS_TRN)
+
+    def _m_inf_T(self, voltage):
+        return 1.0 / (1.0 + exp(-(voltage + 52.0) / 7.4))
+
+    def _h_inf_T(self, voltage):
+        return 1.0 / (1.0 + exp((voltage + 80.0) / 5.0))
+
+    def _tau_h_T(self, voltage):
+        return (85.0 + 1.0 / (exp((voltage + 48.0) / 4.0) + exp(-(voltage + 407.0) / 50.0))) / 3.7371928
```

Comment:
Why are hard-coded constants here?

Reply:
well, the same hardcoded constants are in the original paper and also in the papers that the original paper (Costa et al.) refers to... those are from the experiments in vitro done by Destexhe back in '93 or so... maybe I am not a huge fan of that but in thalamic modelling people use a lot of hard-coded constants, in particular for channel dynamics of a sigmoid type...